cluster-autoscaler/cloudprovider/gce/cache.go

```diff
@@ -147,151 +136,80 @@ func (gc *GceCache) GetMigs() []Mig {
-	klog.V(4).Infof("Regenerating MIG information for %s", migRef.String())
-
-	instances, err := gc.GceService.FetchMigInstances(migRef)
-	if err != nil {
-		klog.V(4).Infof("Failed MIG info request for %s: %v", migRef.String(), err)
-		return err
+	unknown, _ := gc.instancesFromUnknownMig[instanceRef]
+	if unknown {
+		klog.V(5).Infof("Unknown MIG cache hit for %s", instanceRef)
 	}
-
-	return gc.updateInstanceRefToMigRef(migRef, instances)
+	return unknown
 }
 
-func (gc *GceCache) updateInstanceRefToMigRef(migRef GceRef, instances []cloudprovider.Instance) error {
-	// cleanup old entries
-	gc.removeInstancesForMigs(migRef)
+// SetMigInstances assigns MIG ref to the instances refs
+func (gc *GceCache) SetMigInstances(migRef GceRef, instances []cloudprovider.Instance) error {
+	gc.cacheMutex.Lock()
+	defer gc.cacheMutex.Unlock()
 
+	gc.removeMigInstances(migRef)
 	for _, instance := range instances {
 		instanceRef, err := GceRefFromProviderId(instance.Id)
 		if err != nil {
 			return err
 		}
-		gc.instanceRefToMigRef[instanceRef] = migRef
+		gc.instancesToMig[instanceRef] = migRef
 	}
 	return nil
 }
 
-// RegenerateInstancesCache triggers instances cache regeneration under lock.
-func (gc *GceCache) RegenerateInstancesCache() error {
-	gc.instanceRefToMigRef = make(map[GceRef]GceRef)
-	gc.instancesFromUnknownMigs = make(map[GceRef]struct{})
+// MarkInstanceMigUnknown sets instance MIG to unknown
+func (gc *GceCache) MarkInstanceMigUnknown(instanceRef GceRef) {
+	gc.cacheMutex.Lock()
+	defer gc.cacheMutex.Unlock()
 
-	migs := gc.getMigRefs()
-	errors := make([]error, len(migs))
-	ctx, cancel := context.WithCancel(context.Background())
-	workqueue.ParallelizeUntil(ctx, gc.concurrentGceRefreshes, len(migs), func(piece int) {
-		errors[piece] = gc.RegenerateInstanceCacheForMig(migs[piece])
-		if errors[piece] != nil {
-			cancel()
-		}
-	}, workqueue.WithChunkSize(gc.concurrentGceRefreshes))
+	gc.instancesFromUnknownMig[instanceRef] = true
+}
 
-	for _, err := range errors {
-		if err != nil {
-			return err
+// InvalidateInstancesToMig clears the instance to mig mapping for a GceRef
+func (gc *GceCache) InvalidateInstancesToMig(migRef GceRef) {
+	gc.cacheMutex.Lock()
+	defer gc.cacheMutex.Unlock()
+
+	klog.V(5).Infof("Mig instances cache invalidated for %s", migRef)
+	gc.removeMigInstances(migRef)
+}
+
+// InvalidateAllInstancesToMig clears the instance to mig cache
+func (gc *GceCache) InvalidateAllInstancesToMig() {
+	gc.cacheMutex.Lock()
+	defer gc.cacheMutex.Unlock()
+
+	klog.V(5).Infof("Instances to migs cache invalidated")
+	gc.instancesToMig = make(map[GceRef]GceRef)
+	gc.instancesFromUnknownMig = make(map[GceRef]bool)
+}
+
+func (gc *GceCache) removeMigInstances(migRef GceRef) {
+	for instanceRef, instanceMigRef := range gc.instancesToMig {
+		if migRef == instanceMigRef {
+			delete(gc.instancesToMig, instanceRef)
+			delete(gc.instancesFromUnknownMig, instanceRef)
```

Comment:
Agreed, makes perfect sense to keep delete just in case we missed something. I'd consider adding a comment mentioning this so that someone reading the code in future doesn't have to wonder why this is here.